File: nim/anagrams.nim
import std/[strutils, sequtils, cmdline]

type InputArgumentError* = object of IOError

proc generatePermutations(a: var seq[char], n: int) =
  if n <= 0:
    echo a.join("")
  else:
    for i in 0 ..< n:
      generatePermutations(a, n - 1)
      let j = if n mod 2 == 0: 0 else: i
      swap(a[j], a[n])
    generatePermutations(a, n - 1)

when isMainModule:
  var args = commandLineParams()
  if len(args) != 1:
    raise newException(InputArgumentError, "ERROR: There must be exactly 1 argument.")
  var letters = toSeq(args[0].items)
  generatePermutations(letters, len(letters) - 1)

File: csharp/Anagrams.cs
using System;

static class Anagrams
{
    static void GeneratePermutations(char[] a, int n)
    {
        if (n <= 0)
        {
            Console.WriteLine(new string(a));
        }
        else
        {
            for (int i = 0; i < n; i++)
            {
                GeneratePermutations(a, n - 1);
                int j = n % 2 == 0 ? 0 : i;
                (a[j], a[n]) = (a[n], a[j]);
            }
            GeneratePermutations(a, n - 1);
        }
    }

    static void Main(string[] args)
    {
        if (args.Length != 1)
        {
            throw new Exception("ERROR: There must be exactly 1 argument.");
        }
        GeneratePermutations(args[0].ToCharArray(), args[0].Length - 1);
    }
}